src/AutoROM.py

```diff
@@ -129,37 +134,61 @@
     "eea0da9b987d661264cce69a7c13c3bd": "zaxxon",
 }
 
-# simply download tar file to specified dir
-def download_tar_to_buffer(url="https://roms8.s3.us-east-2.amazonaws.com/Roms.tar.gz"):
-    with requests.get(url, stream=True) as response:
-        assert response.status_code == 200
-        assert response.headers["Content-Type"] == "application/x-gzip"
-
-        archive_size = int(response.headers["Content-Length"])
-        assert archive_size > 0
-
-        chunk_size = 2 ** 10
-        with tqdm(
-            unit="B",
-            unit_scale=True,
-            unit_divisor=chunk_size,
-            desc="Downloading ROMs",
-            leave=False,
-            total=archive_size,
-        ) as pbar:
-            buffer = io.BytesIO()
-            for chunk in response.iter_content(chunk_size=chunk_size):
-                buffer.write(chunk)
-                pbar.update(len(chunk))
-
-            buffer.flush()
-            buffer.seek(0)
-            return buffer
+
+class ROMDownloadError(Exception):
+    pass
+
+
+# Download ROMs from a torrent magnet link
+def download_tar_to_tmp_file(timeout_seconds=120):
+    state_str = [
+        'queued',
+        'checking',
+        'downloading metadata',
+        'downloading',
+        'finished',
+        'seeding',
+        'unused enum',
+        'checking fastresume',
+    ]
+
+    ses = lt.session()
+    tmp_dir = tempfile.TemporaryDirectory()
+    params = lt.parse_magnet_uri(MAGNET_LINK)
+    params.save_path = tmp_dir.name
+    handle = ses.add_torrent(params)
+
+    pbar = tqdm(
+        unit="B",
+        unit_scale=True,
+        unit_divisor=1024,
+        desc="Downloading ROMs",
+        leave=False,
+        total=0, # the total download size is only known after downloading metadata
+    )
+    for _ in range(timeout_seconds):
+        status = handle.status()
+        if status.is_seeding:
+            break
+
+        pbar.set_description(state_str[status.state])
+        if status.total_wanted > 0 and status.total != pbar.total:
+            pbar.total = status.total_wanted
+
+        inc = status.total_wanted_done - pbar.n
+        pbar.update(n=inc)
+
+        time.sleep(1)
+    else:
+        raise ROMDownloadError("Download timed out")
+
+    return os.listdir(tmp_dir.name)[0]
```

Comment:
Isn't there some way to list the files you've downloaded from the libtorrent `handle`? It would be nice to parse this from the torrent metadata.

If this isn't too much work you can also just use `pathlib.Path` to be more consistent.

Reply:
:+1: done in https://github.com/Farama-Foundation/AutoROM/pull/16/commits/1820bd661e1ef9bdb68381b479b0e71ceb218997